Stat persistor

kadhikari opened the pull request.

Commit:
jormungandr: Modified for stat

Commit:
jormungandr: Added for stat

Commit:
stat_persistor: service to write stat added

Commit:
jormungandr: StatManager initialized

Commit:
jormungandr: StatManager removed

Commit:
jormungandr: Import removed.

Commit:
jormungandr: StatManager removed from InstanceManager

Commit:
jormungandr: StatManager finalized.

Commit:
stat_persistor: Module to insert statistiques
in the database finalized.

Commit:
jormungandr: Stat_Manager modified

Commit:
stat_persistor: Module to write in database finalized

Commit:
stat_persistor: ForeignKey_Ids managed in sqlAlchemy
instead of database.

Commit:
stat_persistor: Alembic files added for versionning
of stat database.

l-vincent-l commented on source/jormungandr/jormungandr/__init__.py:
```diff
@@ -37,6 +37,10 @@
 
+from jormungandr.stat_manager import StatManager
+i_stat_manager = StatManager()
+i_stat_manager.init("amqp://guest:guest@localhost:5672//","navitia","stat.sender")
```
Il faut que ce soit configurable.

l-vincent-l commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -5,9 +5,10 @@
+import datetime, time
 import base64
 from navitiacommon.models import User, Instance, db
+#from jormungandr.stat_manager import StatManager
```
Cette ligne ne sert plus

l-vincent-l commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -5,9 +5,10 @@
 from jormungandr.exceptions import RegionNotFound
 from jormungandr import i_manager
-import datetime
+import datetime, time
```
Du coup tu ne te sers plus de l'import time ?

l-vincent-l commented on source/jormungandr/jormungandr/interfaces/v1/Journeys.py:
```diff
@@ -3,7 +3,7 @@
 from flask.ext.restful.types import boolean
 from jormungandr import i_manager
-from jormungandr.interfaces.v1.Uri import journey_pattern_points
+#from jormungandr.interfaces.v1.Uri import journey_pattern_points
```
Si ça ne sert plus il faut l'effacer.

l-vincent-l commented on source/jormungandr/jormungandr/scripts/default.py:
```diff
@@ -3,7 +3,7 @@
 import navitiacommon.request_pb2 as request_pb2
 import navitiacommon.response_pb2 as response_pb2
-from jormungandr.interfaces.v1.Journeys import journey
+#from jormungandr.interfaces.v1.Journeys import journey
```
Pareil qu'avant.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+from navitiacommon import stat_pb2
+from navitiacommon import request_pb2
+from datetime import datetime, timedelta
+from datetime import datetime
```
Tu importes deux fois datetime.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            stat_request.user_name = 'Toto' #??
+            stat_request.application_id = 2 #??
+            stat_request.application_name = 'navitia2' #??
+            stat_request.api = request.endpoint
```
je ne suis pas sur que tu veuilles laisser ces variables d'initialisation

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            stat_request.api = request.endpoint
+            stat_request.query_string = request.url
+            stat_request.host = request.host_url
+            stat_request.response_size = 400
```
Pareil si tu n'as pas encore la taille de la réponse, ne l'initialise pas.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            stat_request.host = request.host_url
+            stat_request.response_size = 400
+            stat_request.client = request.remote_addr
+        except Exception:
```
Il faut au moins logger l'erreur. Qu'est-ce qu'il pourrait il y avoir comme erreur ? dans le mktime ?

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            try:
+                stat_parameter.key = item[0]
+                stat_parameter.value = item[1]
+            except Exception:
```
Je pense qu'ici au lieu de faire un try catch, tu peux tester le type de item et sa taille.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+    def fill_coverages(self, stat_request, func_call):
+        stat_coverage = stat_request.coverages.add()
+        try:
+            stat_coverage.region_id = request.view_args['region']
```
Ici il faut faire un 
if "region" in request.view_args

Et enlever le try except.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+        try:
+            self.init_journey(stat_journey)
+            key_name = 'requested_date_time'
+            if key_name in resp_journey:
```
Pourquoi ne pas faire 'requested_date_time' in resp_journey: ?

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            if key_name in resp_journey:
+                req_date_time = resp_journey[key_name]
+                if len(req_date_time) > 0:
+                    stat_journey.requested_date_time = int(time.mktime(datetime.strptime(req_date_time, f_datetime).timetuple()))
```
int(time.mktime(datetime.strptime(req_date_time, f_datetime).timetuple())) mérite une fonction.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            key_name = 'requested_date_time'
+            if key_name in resp_journey:
+                req_date_time = resp_journey[key_name]
+                if len(req_date_time) > 0:
```
tu peux faire directement if req_date_time:

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            if key_name in resp_journey:
+                stat_journey.type = resp_journey[key_name]
+
+        except Exception:
```
S'il y a une exception de levée, il faut du log.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            key_name = 'type'
+            if key_name in resp_section:
+                stat_section.type = resp_section[key_name]
+        except Exception:
```
Mêmes remarques que précédemment.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            if len(to_lon) > 0:
+                stat_section.to_coord.lon = float(to_lon)
+
+        except Exception:
```
Si tu fais un try catch, tu peux t'éviter les if to_lat, en catchant les ValueError, et en loggant.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            stat_section.route_id = self.get_section_link(resp_section, 'route')
+            stat_section.network_id = self.get_section_link(resp_section, 'network')
+            stat_section.physical_mode_id = self.get_section_link(resp_section, 'physical_mode')
+            stat_section.commercial_mode_id = self.get_section_link(resp_section, 'commercial_mode')
```
qu'est-ce qui peut péter ici ?

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+                if (admin['level'] == 8):
+                    stat_section.from_admin_id = admin['id']
+                    stat_section.from_admin_name = admin['name']
+                    pass
```
Je pense que tu voulais dire break

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,301 @@
+            stat_section.network_name = resp_section['display_informations']['network']
+            stat_section.physical_mode_name = resp_section['display_informations']['physical_mode']
+            stat_section.commercial_mode_name = resp_section['display_informations']['commercial_mode']
+        except Exception:
```
Catch le KeyError

l-vincent-l commented on source/stat_persistor/migrations/alembic/env.py:
```diff
@@ -0,0 +1,71 @@
+    connection = engine.connect()
+    context.configure(
+                connection=connection,
+                target_metadata=target_metadata
```
Il faut que tu ajoutes ton schema stat https://groups.google.com/forum/#!topic/sqlalchemy-alembic/uAGOJ4gSHw4 avec include_symbol

l-vincent-l commented on source/stat_persistor/migrations/alembic/versions/176f17297061_create_all_tables.py:
```diff
@@ -0,0 +1,125 @@
+    )
+
+def downgrade():
+    op.drop_table('journey_sections', schema='stat')
```
tu as testé ? des fois y a des problèmes avec les cascade

kadhikari replied:
I tested many times downgrade and upgrade in alembic and it works fine.

l-vincent-l commented on source/stat_persistor/stat_persistor.ini:
```diff
@@ -0,0 +1,24 @@
+#realtime topics used for this intance,it represent the source of realtime information for this ED.
+#it's a list and there always a comma.
+#example: foo, bar
+rt-topics=stat.sender
```
Si tu veux qu'un seul topic, c'est sans s.

antoine-de commented:
test this please

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+    result['month']= from_timestamp(stat.request_date).month
+    result['day']= from_timestamp(stat.request_date).day
+    result['hour']= from_timestamp(stat.request_date).hour
+    result['minute']= from_timestamp(stat.request_date).minute
```
T'aurais peut être pu faire un 
try:
    request_date = from_timestamp(stat.request_date)
except:
    logging.error("Unable to parse date")
    return {}
return {'request_date' : request_date.strftime('%Y%m%d %H:%M:%S'), 
}
....

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+    Construit à partir d'un object protobuf pbnavitia.stat.HitStat.parameters
+    Utilisé pour l'insertion dans la table stat.parameters
+    """
+    result = {}
```
return {
  'request_id' : request_id,
   'param_key': param.key,
   'param_value' : param.value
}

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+    result = {}
+    result['region_id'] = coverage.region_id
+    result['request_id'] = request_id
+    return result
```
Pareil ici

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+    result['duration'] = journey.duration
+    result['nb_transfers'] = journey.nb_transfers
+    result['type'] = journey.type
+    return result
```
Pareil qu'avant

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+    result['commercial_mode_name'] = section.commercial_mode_name
+    result['physical_mode_id'] = section.physical_mode_id
+    result['physical_mode_name'] = section.physical_mode_name
+    return result
```
Idem

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,151 @@
+
+
+
+
```
Pourquoi tant de lignes vides ?

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+class StatSaver(object):
+    """
+    Classe responsable de l'enregistrement en base de donnée des statistiques
+    (dans un fichier pour le moments)
```
C'est valable pour pas mal de commentaire, mais on essaie de les faire en anglais.

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+        self.coverage_table = Table('coverages', self.meta,
+                                    Column('request_id', None,
+                                           ForeignKey('requests.id')),
+                                    autoload=True,
```
Soit tu fais un autoload, soit tu déclares ta table, mais ne fais pas les deux.

http://docs.sqlalchemy.org/en/rel_0_8/core/reflection.html

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+                                           Column('journey_id', None,
+                                                  ForeignKey('journeys.id')),
+                                           autoload=True,
+                                           schema='stat')
```
Le commentaire au dessus est valable pour toutes les tables.

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+                                           schema='stat')
+
+    def persist_stat(self, config, stat_request):
+        self.__persist(stat_request, config, persist_stat_request)
```
Essaie de garder le même ordre des paramètres.

l-vincent-l commented on source/stat_persistor/stat_persistor/daemon.py:
```diff
@@ -0,0 +1,116 @@
+        if stat_hit.IsInitialized():
+            try:
+                self.stat_saver.persist_stat(self.config, stat_hit)
+            except FunctionalError as e:
```
Persist stat raise aussi une technicalerror

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+
+        try:
+            callback(self.meta, conn, item)
+            #callback(item, config)
```
à effacer

kadhikari replied:
Corrected and will be pushed soon.

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+            transaction.rollback()
+            raise FunctionalError(str(e))
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            logger.exception('error durring transaction')
```
Il n'y a qu'un r à durring

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+            logger.exception('error durring transaction')
+            try:
+                transaction.rollback()
+            except:
```
Log si le rollback s'est mal passé.

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/statsaver.py:
```diff
@@ -0,0 +1,100 @@
+                transaction.rollback()
+            except:
+                pass
+            raise
```
Tu raises quoi ?

Commit:
jormungandr: Commented lines deleted.

Commit:
jormungandr: Stat_Manager configuration added.

Commit:
jormungandr: Modifications in Stat_Manager
Error handling
Process of Reading

Commit:
stat_persistor: Items Corrected
Query of insertion.
Error handling

Commit:
jormungandr: Test of initialization
of remote_addr

Commit:
jormungandr: More controls added in stat_manager

antoine-de commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -86,6 +85,7 @@ def authenticate(region, api, abort=False):
 
     user = User.get_from_token(token, datetime.datetime.now())
     if user:
+        #STAT-1 Ajouter flask variable (global) pour pouvoir garder les valeurs
```
c'est un TODO ?
sinon de manière générale vaut mieux mettre les commentaires en anglais

kadhikari replied:
C'est un oublie. J'avais prévu d'ajouter les codes mais finalement je ne l'ai pas fait

antoine-de commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+        self.exchange_name = None
+        self.topic_name = None
+
+    def init(self, save_stat=False, broker_url=None, exchange_name=None, topic_name=None):
```
pourquoi tu mets 2 méthodes (init et '_ _ init _ _' ) ?

kadhikari replied:
Le premier est une méthode de initialisation à appeler alors que l'autre est une méthode constructeur de la class.

kinnou02 replied:
ben justement, ya des cas ou on veux créer la classe sans l'initialiser?

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        Function to fill stat objects (requests, parameters, journeys et sections) sand send them to Broker
+        """
+        if not self.save_stat:
```
Il faut logger ici.

kinnou02 replied:
euh, non, on va pas écrire une ligne de log à chaque requete, et encore moins avec un niveau de log a info

kinnou02 replied:
2éme remarque, je ferais le truc de facon intélligente, pour n'ajouter que si besoin le décorateur de fonction, comme ça, si les stats ne sont pas activé, on ne passe meme du tout dans ces fonctions.

voir plutot qu'un décorateur, on pourrait utiliser [after_request](http://flask.pocoo.org/docs/api/#flask.Flask.after_request)

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+        end_time = time.time()
+        response_duration = (end_time - start_time) #In seconds
+        stat_request = stat_pb2.StatRequest()
+        stat_request.request_duration = int(response_duration * 1000) #In milliseconds
```
fais directement le int((end_time - start_time)*1000)

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+        stat_request.user_name = ''
+        stat_request.application_id = -1
+        stat_request.application_name = ''
+        stat_request.api = request.endpoint
```
Il faudrait peut être avoir une map des endpoints vers les api.
On pourrait aussi logger la version (ça c'est à voir avec chef).

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+    def fill_parameters(self, stat_request, func_call):
+        for item in request.args.iteritems():
+            stat_parameter = stat_request.parameters.add()
+            if len(item) > 1:
```
l'inverse ne peut pas arriver.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+
+    def fill_coverages(self, stat_request, func_call):
+        stat_coverage = stat_request.coverages.add()
+        if 'region' in request.view_args:
```
Tu loupes tous les journeys.
Je pense que tu peux passer la région aussi dans flask.g

kinnou02 replied:
je seconde

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+
+        if 'from' in resp_section:
+            section_from = resp_section['from']
+            if 'embedded_type' in section_from:
```
tu l'as nécessairement.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+            if 'embedded_type' in section_from:
+                stat_section.from_embedded_type = section_from['embedded_type']
+
+            if 'id' in section_from:
```
pareil

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+                stat_section.from_coord.lon = float(from_lon)
+
+            except ValueError as e:
+                    logging.getLogger(__name__).warn("%s", str(e))
```
Pour les logs dans flask  : 
http://flask.pocoo.org/docs/api/#flask.Flask.logger

Il faut récupérer le contexte d'application : 
from flask import current_app

Puis pour logger: current_app.logger.warn(...)

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+                        stat_section.from_admin_name = admin['name']
+                        break
+
+            if 'administrative_regions' in resp_section['to'][stat_section.from_embedded_type]:
```
le from et le to n'ont pas nécéssairement les mêmes types.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+                        break
+
+        except ValueError as e:
+            logging.getLogger(__name__).warn("%s", str(e))
```
Juste pour te rappeler de changer le log.

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+            self.manager.manage_stat(start_time, func_call)
+            return func_call
+        return wrapper
+
```
ça serait cool d'enlever les lignes de fin.

l-vincent-l commented on source/stat_persistor/stat_persistor.ini:
```diff
@@ -0,0 +1,24 @@
+
+#realtime topics used for this intance,it represent the source of realtime information for this ED.
+#it's a list and there always a comma.
+#example: foo, bar
```
Du coup si ce n'est pas une liste, le commentaire est faux.

l-vincent-l commented on source/stat_persistor/stat_persistor/config.py:
```diff
@@ -0,0 +1,75 @@
+
+        confspec.append('[stat_persistor]')
+        confspec.append('exchange-name = string(default="navitia")')
+        confspec.append('rt-topic = string(default="stat.sender")')
```
Heu, du coup ce n'est pas du temps réel si ?

l-vincent-l commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,307 @@
+        Remplir requests
+        """
+        dt = datetime.now()
+        stat_request.request_date = int(time.mktime(dt.timetuple()))
```
Y a  un intérêt à ne pas mettre directement dans le bon format ?

kadhikari replied:
C'est pour séparer l'objet stat de jormungandr/stat_persistor et la base de données.

kinnou02 replied:
Deux remarques: 
- si timestamp: ça doit être l'heure UTC
- ca tombe bien, on ne doit plus que travailler avec des heures UTC, et now retourne une date local

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,159 @@
+    Inserer les statistiques pour chaque requete:
+    stat.request, stat.parameters, stat.journeys, stat.journey_sections
+    """
+    request_table = meta.tables['stat.requests']
```
il faut préciser dans l'object metadata le schema.

kinnou02 replied:
ça aurait probablement était plus simple d'utiliser la partie ORM de sqlalchemy

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,159 @@
+        'month': from_timestamp(stat.request_date).month,
+        'day': from_timestamp(stat.request_date).day,
+        'hour': from_timestamp(stat.request_date).hour,
+        'minute': from_timestamp(stat.request_date).minute,
```
On n'a pas déjà ces information dans request_date ?

kadhikari replied:
Si. On ajoute ces informations pour accélérer les consultations et pouvoir regrouper par année, mois....

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,159 @@
+        'hour': from_timestamp(stat.request_date).hour,
+        'minute': from_timestamp(stat.request_date).minute,
+        'user_id': stat.user_id,
+        'user_name': stat.user_name,
```
Si on a les id on devrait pouvoir retrouver les noms.

kadhikari replied:
On ne veut pas utiliser une table de référence. Toutes informations sont dans la base stat.

l-vincent-l commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,159 @@
+def get_datetime_from_timestamp(date_time):
+    try:
+        request_date_time = from_timestamp(date_time)
+    except:
```
https://docs.python.org/2/library/datetime.html ça ne devrait lever qu'une valueerror

Commit:
jormungandr: global paramter g added.

Commit:
jormungandr: Corrections in stat_manager

Commit:
stat_persistor: some corrections in stat_persistor

kinnou02 commented on source/jormungandr/jormungandr/__init__.py:
```diff
@@ -37,6 +37,13 @@
+i_stat_manager.init(save_stat=app.config['SAVE_STAT'],
+                    broker_url=app.config['BROKER_URL'],
+                    exchange_name=app.config['EXCHANGE_NAME'],
+                    topic_name=app.config['TOPIC_NAME'])
```
il n'y a pas de notion de topic vu qu'on ne va pas utiliser un exchange de type "topic"

kinnou02 commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -67,6 +66,8 @@ def get_token():
 
 
 def authenticate(region, api, abort=False):
+    #Here we initialize user
```
l'interet d'un commentaire, c'est d'aider à la compréhension, là il ne sert a rien

kinnou02 commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -85,8 +86,11 @@ def authenticate(region, api, abort=False):
     if user:
+        #Here we keep user_id to be used in stat
         if user.has_access(region, api):
+            g.user = {"user_id": user.id}
```
pourquoi ne pas directement stocker l'objet user complet?

kinnou02 commented on source/jormungandr/jormungandr/default_settings.py:
```diff
@@ -55,3 +55,11 @@
+#Paramètres pour les statistiques
+SAVE_STAT = True
+BROKER_URL = 'amqp://guest:guest@localhost:5672//'
+EXCHANGE_NAME='navitia'
```
ca sera un autre exchange que NAVITIA, grosso modo l'exhange NAVITIA correspond aux données de prod, et chaque composant s'abonne sur ce dont il a besoin.

Là le but est plutot de faire une queue de traitement, et si on a plusieurs instances de stat persistor, ils doivent se partager le travail.

kinnou02 commented on source/jormungandr/jormungandr/__init__.py:
```diff
@@ -37,6 +37,13 @@
 i_manager.initialisation(start_ping=app.config['START_MONITORING_THREAD'])
 
+from jormungandr.stat_manager import StatManager
+i_stat_manager = StatManager()
```
pourquoi "i_stat_manager"?

kinnou02 commented on source/jormungandr/jormungandr/interfaces/v1/ResourceUri.py:
```diff
@@ -22,6 +24,7 @@ def __init__(self, *args, **kwargs):
         self.method_decorators.append(add_pagination_links())
         self.method_decorators.append(clean_links())
         self.method_decorators.append(authentification_required)
+        self.method_decorators.append(manage_stat_caller(i_stat_manager))
```
ça pourrai valoir le coup de sous classer Ressource pour avec un truc commun à tout les appels

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+        connection to rabbitmq and initialize queues
+        """
+        self.connection = kombu.Connection(self.broker_url)
+        exchange = kombu.Exchange(self.exchange_name, 'topic', durable = True)
```
on veut un exchange de type "direct"

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+
+    def param_exists(self, param_name, stat_request):
+        result = False
+        for param in stat_request.parameters:
```
ce qui revient à faire:

```
return param_name in parameters
```

et vu que tu n'utilise que parameters de l'obect stat_request, pas besoin de passer tout stat_request

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+            stat_request.user_id = g.user['user_id']
+        if g.user is not None and 'user_name' in g.user:
+            stat_request.user_name = g.user['user_id']
+        stat_request.application_id = -1
```
c'est peut etre pas la peine de le remplir si on a aps la valeur

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+
+        #We do not save informations of journeys and sections for a request
+        #Isochron (parameter "&to" is absent for API Journeys )
+        if 'journeys' in request.endpoint and \
```
endpoint est une collection?

kadhikari replied:
Non, c'est une chaine qui contient la valeur d'api : exemple v1.journeys

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+        stat_request.application_id = -1
+        stat_request.application_name = ''
+        stat_request.api = request.endpoint
+        stat_request.query_string = request.url
```
request.url ne contient pas que la query_string

kadhikari replied:
Peut-on utiliser  "request.query_strin" qui contient que la partie requête (par exemple : "q=cerdon&type[]=stop_area" dans l'appel "../places?")?

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+        stat_request.query_string = request.url
+        stat_request.host = request.host_url
+        if request.remote_addr:
+            stat_request.client = request.remote_addr
```
faut vérifier qu'il tient compte de [x-forward-for](https://en.wikipedia.org/wiki/X-Forwarded-For)

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+        stat_request.response_size = sys.getsizeof(func_call[0])
+
+
+    def fill_parameters(self, stat_request, func_call):
```
func_call n'est pas utilisé

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+            stat_coverage.region_id = request.view_args['region']
+
+
+    def init_journey(self, stat_journey):
```
pas besoin que ce soit une methode, une fonction libre ferait tout aussi bien l'affaire

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+
+    def fill_journey(self, stat_journey, resp_journey):
+        """"
+        Fill journey and all sections:
```
ca serai sympa de dire ce qu'il y a dans resp_journeys

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+            stat_journey.type = resp_journey['type']
+
+
+    def get_posix_date_time(self, date_time_value):
```
pareil c'est une fonction libre

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+            try:
+                return int(time.mktime(datetime.strptime(date_time_value, f_datetime).timetuple()))
+            except ValueError as e:
+                logging.getLogger(__name__).warn("%s", str(e))
```
suis pas persuadé que le try/except soit vraiment requis, mais dans tous les cas il faut préférer:

```
logging.exception
```

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+            self.fill_sections(stat_journey, resp_journey)
+
+
+    def get_section_link(self, resp_section,link_type):
```
espace :@

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+
+
+    def fill_sections(self, stat_journey, resp_journey):
+
```
une ligne vide dés le début s'pas beau

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,295 @@
+        try:
+            if 'administrative_regions' in resp_section['from'][stat_section.from_embedded_type]:
+                for admin in resp_section['from'][stat_section.from_embedded_type]['administrative_regions']:
+                    if (admin['level'] == 8):
```
pas de parenthèse en python

kinnou02 commented on source/stat_persistor/stat_persistor.ini:
```diff
@@ -0,0 +1,22 @@
+[stat]
```
on avait dit qu'on utiliserais un fichier de conf en json, s'plus simple a parser, et beaucoup plus puissant, surtout pour la config des logs

kinnou02 replied:
donc on peux supprimer ce fichier

kinnou02 commented on source/stat_persistor/stat_persistor/daemon.py:
```diff
@@ -0,0 +1,119 @@
+        queue_name = self.config.queue_name + '_' + self.config.rt_topic
+        queue = kombu.Queue(queue_name, exchange=exchange, durable=True,
+            routing_key=self.config.rt_topic)
+        self.queues.append(queue)
```
vu qu'il n'y aura toujours qu'une seule Queue, pas besoin d'avoir une liste de queues

kinnou02 commented on source/stat_persistor/stat_persistor/daemon.py:
```diff
@@ -0,0 +1,119 @@
+            self.handle_statistique(stat_request)
+
+            message.ack()
+        except TechnicalError:
```
Il faut revoir la gestion d'erreur, typiquement on ne passera jamais dans ce except...

kinnou02 replied:
ça aussi

kinnou02 commented on source/type/stat.proto:
```diff
@@ -0,0 +1,77 @@
+    optional string value = 2;    
+}
+
+message Coord{
```
pourquoi ne pas utiliser "GeographicalCoord"?

kinnou02 commented on source/type/stat.proto:
```diff
@@ -0,0 +1,77 @@
+    optional int32	response_size = 11;
+    optional string	client = 12;
+    repeated StatCoverage coverages = 13;    
+    repeated StatParameter parameters = 14;
```
c'est un peu redondant avec la query string

kinnou02 commented on source/type/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 find_package(Protobuf REQUIRED)
-PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS type.proto response.proto request.proto task.proto realtime.proto)
+PROTOBUF_GENERATE_CPP(PROTO_SRCS PROTO_HDRS type.proto response.proto request.proto task.proto realtime.proto stat.proto)
```
je ne suis pas persuadé que ce soit utile de générer le cpp, vu qu'on ne s'en sert null par

kinnou02 commented:
Et il faut absolument des TU, surtout sur la partie jormungandr, le moindre changement sur les interfaces et ça ne marche plus, voir çà fait planter la restitution

Commit:
jormungandr: Abstract class  for statistics

Abstract class to handle statistics on all queries

Commit:
jormungandr: corrections for statistics

Commit:
stat_persistor: Corrections after review

Commit:
type: correction for statistic

Commit:
stat: add license

kinnou02 commented on source/jormungandr/jormungandr/authentification.py:
```diff
@@ -124,9 +125,12 @@ def authenticate(region, api, abort=False):
         else:
             return False if not instance else instance.first().is_free
 
-    user = get_user()
```
tu es en train d'annuler l'ensemble des changements que j'ai fait sur l'authentification là...

kinnou02 commented on source/jormungandr/jormungandr/default_settings.py:
```diff
@@ -55,3 +55,10 @@
 }
+
+#Paramètres pour les statistiques
+SAVE_STAT = True
```
pour uen conf par défaut il vaut mieux mettre false

kinnou02 commented on source/jormungandr/jormungandr/interfaces/v1/StatedResource.py:
```diff
@@ -0,0 +1,42 @@
+        Resource.__init__(self, *args, **kwargs)
+        self.method_decorators = []
+
+        if True:#stat_actived:
```
c'est un peu radical

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,376 @@
+        if not self.save_stat:
+            return
+
+        #start_time = time.time();
```
si on en a pas besoin on peux le supprimer

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,376 @@
+        self.publish_request(stat_request)
+
+
+    def fill_Result(self, stat_request, func_call):
```
pas de majuscule sur les noms de fonction

kinnou02 commented on source/jormungandr/jormungandr/stat_manager.py:
```diff
@@ -0,0 +1,376 @@
+        """
+        dt = datetime.utcnow()
+        stat_request.request_date = int(time.mktime(dt.timetuple()))
+        if hasattr(g,'user') and g.user is not None:
```
tu peut directement utiliser la fonction get_user du module authentification

kinnou02 commented on source/stat_persistor/stat_persistor.json:
```diff
@@ -0,0 +1,12 @@
+       "exchange-name":"stat_persistor_exchange",
+       "queue-name":"stat_persistor_queue",
+       "broker-url":"amqp://guest:guest@localhost:5672//",
+       "log-level":"debug",
```
il faut changer la conf du logger pour utiliser: https://docs.python.org/dev/library/logging.config.html#logging.config.dictConfig

kinnou02 replied:
C'est toujours d'actualité

kinnou02 commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,202 @@
+# encoding: utf-8
```
pas de majuscule dans les nom de fichier

kinnou02 commented on source/stat_persistor/stat_persistor/saver/stat_Request.py:
```diff
@@ -0,0 +1,202 @@
+    Construit à partir d'un object protobuf pbnavitia.stat.HitStat.journey.sections
+    Utilisé pour l'insertion dans la table stat.sections
+    """
+    from_point = "POINT(%.20f %.20f)" %(section.from_coord.lon, section.from_coord.lat)
```
il faut privilégier "format()"

Commit:
stat: stat_Request file renamed.

Commit:
stat: file stat_persistor.ini deleted.

Commit:
jormungandr: Corrections after review.

Commit:
jormungandr test: paramter save_stat disabled

For the moment we don't test authentication in the integration test
not to have dependency to the jormungandr database.

Commit:
tests: test units for stat added.

Commit:
stat: corrections after remarks.

Commit:
jormungandr: methods displaced.

Methods displaced from routing_tests.py to check_utils.py

Commit:
test: Modified to adjust tests on stat.

Commit:
jormungandr: SAVE_STAT deactivated.

Commit:
test: correction in the test for stat.

Commit:
jormungandr: stat test case added for place_uri.

Commit:
jormungandr: Correction after remarks

Commit:
jormungandr: corrections on stat test.

kinnou02 commented on source/jormungandr/tests/no_stats_tests.py:
```diff
@@ -0,0 +1,79 @@
+
+    def test_simple_journey_query_stats(self):
+        """
+        If the stat manager is failing and we want stats, we get an error
```
ben j'doit dire que je suis pas top d'accord, ca serait cool si ca pouvais ne pas foutre en l'air le service

kinnou02 commented:
Note pour plus tard: faudra ajouter de la supervision sur le service

antoine-de committed:
Integration tests: no connection to jormun db

mock get_from_token method in integration tests to skip the connection
to the jormun database

antoine-de commented:
test this please

Commit:
jormungandr: Error catched in stat manager
In case of error in stat manager jormungandr must work.

Commit:
stat_persistor: corrected after remarks

Commit:
stat_persistor: corrected for remarks of Vincent

Commit:
jormungandr: unnecessary comments deleted.

Commit:
jormungandr: Exception type changed.

Commit:
stat_persistor: message queue consumption

In case of error we retry again without consuming the message.

Commit:
stat_persistor: corrections in error handling
